text/2019-03-19-calculated-commit-ts.md

```diff
@@ -0,0 +1,190 @@
+prewrite, which means, Txn2's `commit_ts` must be larger than Txn1's `start_ts`.
+
+On the other hand, `commit_ts` can't be arbitrarily large. If the `commit_ts` is
+ahead of actual time, the committed data may be unreadable by other new
+transactions, which breaks the integrity. We are not sure whether a timestamp is
+ahead of the real time if we don't ask PD.
+
+To conclude, in order not to break the Snapshot Isolation and the integrity, a
+valid range for `commit_ts` should be:
+
+```text
+max{start_ts, max_read_ts_of_written_keys} < commit_ts <= now
+```
+
+So here comes an idea that we can calculate the commit_ts this way:
+
+```text
+commit_ts = max{start_ts, region_1_max_read_ts, region_2_max_read_ts, ...} + 1
+```
+
+where `region_N_max_read_ts` is the maximum timestamp of all reads on the
+region, for all regions involved in the transaction.
+
+### Code changes to TiKV
+
+A new component named `ReadTsCache` will be added to `Storage` and `Coprocessor`
+(and they share the same content). `ReadTsCache` records `max_read_ts` of all
+leaders on this store. For each leader, these fields are saved:
+
+```rust
+struct RegionMaxTsRecord {
+    /// The `max_read_ts` is not safe to use immediately after creation or
+    /// updating of the region. When the `max_read_ts` is ready, `is_ready` will
+    /// be set.
+    is_ready: bool,
+
+    /// The version of the region's epoch. The `max_read_ts` can only be used
+    /// when `version` matches the prewrite request's version.
+    version: u64,
+
+    /// The maximum timestamp of transactional reads on this region. Note that
+    /// in order to ensure the `max_read_ts` is no less than reads on the region
+    /// before the region's leader goes to the current store, A timestamp will
+    /// be fetched from PD after creating or updating of the region.
+    max_read_ts: u64,
+}
+```
+
+`ReadTsCache` contains mapping from `region_id` to `RegionMaxTsRecord`:
+
+```rust
+// Divide the map into multiple according to region_id's hash to reduce lock
+// contention.
+max_read_ts_map: Vec<RwLock<HashMap<u64, Arc<RwLock<RegionMaxTsRecord>>>>>,
+```
+
+Each transactional read needs to update the `max_read_ts` of the requested
+region **before** getting snapshot (`u64::MAX` will be ignored). Each prewrite
+needs to get the `max_read_ts` after writing successfully and return it with
+`PrewriteResponse`, and the TiKV client is responsible for calculating the
+`commit_ts` according to the timestamps carried in `PrewriteResponse`s. If TiKV
+can't provide `max_read_ts` for a Region (due to not ready or version mismatch),
+just returns 0 to tell the client to fall back to allocating `commit_ts` from
+PD.
+
+`ReadTsCache` updates its inner region information via Raftstore observer and
+only collects leader information. Each time a region is added to the map or a
+region's version is changed, `is_ready` will be unset until getting a timestamp
```

Comment:
"getting a timestamp" is not an act of `is_ready`, so we need to use the passive voice here
```suggestion
region's version is changed, `is_ready` will be unset until a timestamp is obtained
```